### PR TITLE
Reworks how AI tracking is handled & reorganizes it

### LINKED ISSUE #42355: ling's digital camouflage can be countered by AI's Track Person
If AI starts tracking a ling before he uses digital camo AI eye will follow the ling, making the camo useless.

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

Completely reworks how AI tracking is handled, this has no in-game effects.
This moves nearly all AI tracking handling onto ``/datum/tracking``, which previously was pretty bad. I tried documenting as much as I can, making comments actually useful and give accurate information.

Turns ``get_camera_list`` into a global proc, which we now use for camera consoles (including the app), cutting down on copy paste in 2 areas and standardizing its behavior/backend.

## Why It's Good For The Game

I wanted to generalize this behavior so I can use it for tracking players in https://github.com/tgstation/tgstation/pull/77713 - This helps me do that.
Also the current state of AI tracking code is pretty poor and hopefully this improves that area.

Closes https://github.com/tgstation/tgstation/issues/42355

## Changelog

:cl:
refactor: AI's player-tracking eyes received an unwanted obligatory update, and should now not tell you that a player is untrackable when they clearly obviously can be.
/:cl:

### CODE REVIEW

**Comment by Sealed101** [comment]
have some issues
- #42355
- #36526

surely you could fix those while you're in the general area right? i think there's more though, will check back if i find em

**Comment by JohnFulpWillard** [comment]
changeling pov
image
image

ai pov
image
image
image

File: code/__HELPERS/cameras.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/proc/near_camera(mob/living/mob_to_check)
+	if (!isturf(mob_to_check.loc))
+		return FALSE
+	if(issilicon(mob_to_check))
+		var/mob/living/silicon/silicon_tracked = mob_to_check
+		if((QDELETED(silicon_tracked.builtInCamera) || !silicon_tracked.builtInCamera.can_use()) && !GLOB.cameranet.checkCameraVis(silicon_tracked))
+			return FALSE
+	else if(!GLOB.cameranet.checkCameraVis(mob_to_check))
+		return FALSE
+	return TRUE
```
Comment (Jacquerel) [inline code/__HELPERS/cameras.dm:11]:
```suggestion
	if(issilicon(mob_to_check))
		var/mob/living/silicon/silicon_tracked = mob_to_check
		if(silicon_tracked.builtInCamera?.can_use())
			return TRUE
	return GLOB.cameranet.checkCameraVis(mob_to_check)
```
Does this do the same thing?

File: code/__HELPERS/cameras.dm [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+/proc/camera_sort(list/L)
```
Comment (Jacquerel) [inline code/__HELPERS/cameras.dm:13]: Please document this proc, I don't think it is necessarily obvious what it is doing. 
(It seems like it is only called by `get_camera_list` and will never need to be used anywhere else but it's still good for future people touching this code).
Also if there's a better variable name than `L` then use that.

File: code/game/machinery/camera/trackable.dm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			to_chat(tracking_holder, span_warning("Target is not near any active cameras. Attempting to reacquire..."))
+		cameraticks++
+		return
+
+	to_chat(tracking_holder, span_warning("Unable to reacquire, cancelling track..."))
+	cameraticks = initial(cameraticks)
+	set_tracking(FALSE)
+
+///Generates a list of trackable people by name, returning a list of Humans + Non-Humans that can be tracked.
+/datum/trackable/proc/trackable_mobs()
```
Comment (Jacquerel) [inline code/game/machinery/camera/trackable.dm:57]: Picky but can we call this `find_trackable_mobs` instead? `trackable_mobs` sounds like a var name.

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Nitpick, mostly it just took me a few minutes of staring to figure out what `camera_sort` was for and I think we can avoid that with documentation.

File: code/__HELPERS/cameras.dm (hunk 2 above)
Comment (JohnFulpWillard) [reply code/__HELPERS/cameras.dm:13]: This was just moved over

Comment (Jacquerel) [reply code/__HELPERS/cameras.dm:13]: Sure, but you did touch it and you're going to be making more changes in this file anyway 😅

File: code/__HELPERS/cameras.dm (hunk 1 above)
Comment (JohnFulpWillard) [reply code/__HELPERS/cameras.dm:11]: yeah it seems fine

**Pull request merged**